crates/pallet-domains/src/staking.rs

```diff
@@ -0,0 +1,366 @@
+use sp_runtime::traits::{CheckedAdd, Zero};
+use sp_runtime::Percent;
+use sp_std::vec::Vec;
+
+/// Type that represents an operator pool details.
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct OperatorPool<Balance> {
+    pub signing_key: ExecutorPublicKey,
+    pub current_domain_id: DomainId,
+    pub minimum_nominator_stake: Balance,
+    pub nomination_tax: Percent,
+    /// Total active stake for the current pool.
+    pub current_total_stake: Balance,
+    /// Total stake for the current pool in the next epoch.
+    pub next_total_stake: Balance,
+    /// Total shares of the nominators and the operator in this pool.
+    pub total_shares: Balance,
+    pub is_frozen: bool,
+}
+
+/// Type that represents a nominator's details under a specific operator pool
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct Nominator<Balance> {
+    pub shares: Balance,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub enum Transfer<Balance> {
+    Withdraw(Balance),
+    Deposit(Balance),
+}
+
+/// Type that represents a pending transfer
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct PendingTransfer<NominatorId, Balance> {
+    pub nominator_id: NominatorId,
+    pub transfer: Transfer<Balance>,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct StakingSummary<OperatorId, Balance> {
+    /// Current epoch index for the domain.
+    pub current_epoch_index: EpochIndex,
+    /// Total active stake for the current epoch.
+    pub current_total_stake: Balance,
+    /// Total stake for the next epoch.
+    pub next_total_stake: Balance,
+    /// Current operators for this epoch
+    pub current_operators: Vec<OperatorId>,
+    /// Operators for the next epoch.
+    pub next_operators: Vec<OperatorId>,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct OperatorConfig<Balance> {
+    pub signing_key: ExecutorPublicKey,
+    pub minimum_nominator_stake: Balance,
+    pub nomination_tax: Percent,
+}
+
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    MaximumOperatorId,
+    DomainNotInitialized,
+    InsufficientBalance,
+    BalanceFreeze,
+    MinimumOperatorStake,
+    UnknownOperator,
+    MinimumNominatorStake,
+    BalanceOverflow,
+}
+
+pub(crate) fn do_register_operator<T: Config>(
+    operator_owner: T::AccountId,
+    domain_id: DomainId,
+    amount: BalanceOf<T>,
+    config: OperatorConfig<BalanceOf<T>>,
+) -> Result<OperatorId, Error> {
+    DomainStakingSummary::<T>::try_mutate(domain_id, |maybe_domain_stake_summary| {
+        let operator_id = NextOperatorId::<T>::get();
+        let next_operator_id = operator_id.checked_add(1).ok_or(Error::MaximumOperatorId)?;
+        NextOperatorId::<T>::set(next_operator_id);
+
+        OperatorIdOwner::<T>::insert(operator_id, operator_owner.clone());
```

Comment:
If I understand correctly, the current design allows only one domain per operator account id, but it's unchecked whether the operator has already staked on some domain.

Reply:
IIUC, This is not true since the freeze is set on the operator_id. So if I want to run two operators with 100 each, the a total of 200 is frozen with 100 associated with operator 1 and other with operator 2.

Reply:
I see, it's possible that one account id can correspond to multiple operator_ids, each of which may point to the same or different domains, so there is a chance that one account id has multiple operator identities on the same domain (call `register_operator` again by accident?).

As a user (owning the private key of some account_id), I care more about what domains this account id corresponds to, not the operator_id, should we add a constraint that an account id can only have one operator_id on a certain domain? So that it can be avoided that there are possibly multiple operator entries under one domain in users' wallet, which is weird IMHO.

Reply:
> I see, it's possible that one account id can correspond to multiple operator_ids, each of which may point to the same or different domains, so there is a chance that one account id has multiple operator identities on the same domain (call register_operator again by accident?).

Yes absolutely possible.

> As a user (owning the private key of some account_id), I care more about what domains this account id corresponds to, not the operator_id, should we add a constraint that an account id can only have one operator_id on a certain domain? So that it can be avoided that there are possibly multiple operator entries under one domain in users' wallet, which is weird IMHO.

Assuming we implement this constraint per domain
- Imagine the scenario where I as an operator operates domain#1 and domain#2. At some point, I realize its profitable to switch to domain#1 from domain#2, if we have a constraint you have proposed, then this is not possible directly. In this case, I first have to de-register from Domain#2 and then register to Domain#1 but its much slower than directly switching which is just one epoch at max. 
- There would another storage item doing a reverse mapping of account_id => operator_id for each domain.

Ofcourse, we can make operator_id and account_id a bijection globally but this leads to managing each account per operator and an extra reverse storage mapping.

My main concern is, given the current implementation, would there be a potential attack scenario where a single operator_account with multiple operator_ids under the same domain?
For example: running single domain client with same operator config but different operator_id ?